command/013_config_upgrade.go

```diff
@@ -71,58 +75,386 @@ func (c *ZeroThirteenUpgradeCommand) Run(args []string) int {
+			out = hclwrite.NewEmptyFile()
+		} else if err != nil {
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				"Unable to read configuration file",
+				fmt.Sprintf("Error when reading configuration file %q: %s", filename, err),
+			))
+			c.showDiagnostics(diags)
+			return 1
+		} else {
+			// Configuration file already exists, so load and parse it
+			config, err := ioutil.ReadFile(filename)
+			if err != nil {
+				diags = diags.Append(tfdiags.Sourceless(
+					tfdiags.Error,
+					"Unable to read configuration file",
+					fmt.Sprintf("Error when reading configuration file %q: %s", filename, err),
+				))
+				c.showDiagnostics(diags)
+				return 1
+			}
+			var parseDiags hcl.Diagnostics
+			out, parseDiags = hclwrite.ParseConfig(config, filename, hcl.InitialPos)
+			diags = diags.Append(parseDiags)
+		}
+
+		if diags.HasErrors() {
+			c.showDiagnostics(diags)
+			return 1
+		}
+
+		// Find all required_providers blocks, and store them alongside a map
+		// back to the parent terraform block.
+		var requiredProviderBlocks []*hclwrite.Block
+		parentBlocks := make(map[*hclwrite.Block]*hclwrite.Block)
+		root := out.Body()
+		for _, rootBlock := range root.Blocks() {
+			if rootBlock.Type() != "terraform" {
+				continue
+			}
+			for _, childBlock := range rootBlock.Body().Blocks() {
+				if childBlock.Type() == "required_providers" {
+					requiredProviderBlocks = append(requiredProviderBlocks, childBlock)
+					parentBlocks[childBlock] = rootBlock
+				}
+			}
+		}
+
+		// First required provider block, and the rest found in this file.
+		var first *hclwrite.Block
+		var rest []*hclwrite.Block
+
+		if len(requiredProviderBlocks) > 0 {
+			// If we already have one or more required provider blocks, we'll rewrite
+			// the first one, and remove the rest.
+			first, rest = requiredProviderBlocks[0], requiredProviderBlocks[1:]
+		} else {
+			// Otherwise, find or a create a terraform block, and add a new
+			// empty required providers block to it.
+			var tfBlock *hclwrite.Block
+			for _, rootBlock := range root.Blocks() {
+				if rootBlock.Type() == "terraform" {
+					tfBlock = rootBlock
+					break
+				}
+			}
+			if tfBlock == nil {
+				tfBlock = root.AppendNewBlock("terraform", nil)
+			}
+			first = tfBlock.Body().AppendNewBlock("required_providers", nil)
+		}
+
+		// Find the body of the first block to prepare for rewriting it
+		body := first.Body()
+
+		// Build a sorted list of provider local names, for consistent ordering
+		var localNames []string
+		for localName := range requiredProviders {
+			localNames = append(localNames, localName)
+		}
+		sort.Strings(localNames)
+
+		// Populate the required providers block
+		for _, localName := range localNames {
+			requiredProvider := requiredProviders[localName]
+			var attributes = make(map[string]cty.Value)
+
+			if !requiredProvider.Type.IsZero() {
+				attributes["source"] = cty.StringVal(requiredProvider.Type.String())
```

Comment:
I think I'd rather generate this as `requiredProvider.Type.ForDisplay()` here so that the `registry.terraform.io/` prefix will be omitted in the common case. I expect a significant number of users can get away with never learning that other registry hosts are possible, because most will only be using the HashiCorp-maintained and partner ones, and I think it'll be slightly easier for them to quickly understand what's going on if e.g. `aws` just becomes `"hashicorp/aws"`.

I noticed that the comment it'll generate for an unknown provider includes an example with a hostname, which I like because it means that we can give the hint of that possibility to the folks most likely to benefit from it while keeping the cognitive load of reviewing the proposed upgrade smaller for everyone else.

Reply:
Good call! Done.